kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerManager.java

```diff
@@ -78,34 +94,52 @@
   // during read operations.
   private final Map<ConsumerInstanceId, KafkaConsumerState> consumers =
       new HashMap<ConsumerInstanceId, KafkaConsumerState>();
-  // Read operations are common and there may be many concurrently, so they are farmed out to
-  // worker threads that can efficiently interleave the operations. Currently we're just using a
-  // simple round-robin scheduler.
-  private final List<KafkaConsumerWorker> workers;
-  private final AtomicInteger nextWorker;
-  // A few other operations, like commit offsets and closing a consumer can't be interleaved, but
-  // they're also comparatively rare. These are executed serially in a dedicated thread.
+  // All kind of operations, like reading records, committing offsets and closing a consumer
+  // are executed separately in dedicated threads via a cached thread pool.
   private final ExecutorService executor;
   private KafkaConsumerFactory consumerFactory;
   private final PriorityQueue<KafkaConsumerState> consumersByExpiration =
       new PriorityQueue<KafkaConsumerState>();
+  private final DelayQueue<RunnableReadTask> delayedReadTasks = new DelayQueue<>();
   private final ExpirationThread expirationThread;
+  private ReadTaskSchedulerThread readTaskSchedulerThread;
 
-  public KafkaConsumerManager(KafkaRestConfig config) {
+  public KafkaConsumerManager(final KafkaRestConfig config) {
     this.config = config;
     this.time = config.getTime();
     this.bootstrapServers = config.bootstrapBrokers();
-    this.workers = new Vector<KafkaConsumerWorker>();
-    for (int i = 0; i < config.getInt(KafkaRestConfig.CONSUMER_THREADS_CONFIG); i++) {
-      KafkaConsumerWorker worker = new KafkaConsumerWorker(config);
-      workers.add(worker);
-      worker.start();
-    }
-    nextWorker = new AtomicInteger(0);
-    this.executor = Executors.newFixedThreadPool(1);
+
+    // Cached thread pool
+    int maxThreadCount = config.getInt(CONSUMER_MAX_THREADS_CONFIG) < 0 ? Integer.MAX_VALUE
+            : config.getInt(CONSUMER_MAX_THREADS_CONFIG);
+
+    this.executor = new ThreadPoolExecutor(0, maxThreadCount,
+            60L, TimeUnit.SECONDS,
+            new SynchronousQueue<Runnable>(),
+            new RejectedExecutionHandler() {
+            @Override
+            public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+              log.debug("The runnable {} was rejected execution."
+                  + "The thread pool must be satured or shutiing down", r);
+              if (r instanceof RunnableReadTask) {
+                RunnableReadTask readTask = (RunnableReadTask) r;
+                int delayMs = ThreadLocalRandom.current().nextInt(25, 75 + 1);
+                readTask.waitExpirationMs = config.getTime().milliseconds() + delayMs;
```

Comment:
I am wondering if it makes sense to add a `RunnableReadTask.delay(long delayMs)`, which is used both here and when we are backing off to set `waitExpirationMs` and add to the delay queue. Then we could add the logic to check for expiration of the task in that method so that we have it down both paths.

Reply:
That certainly is a better abstraction.
What task expiration are you referring to?